## Errors and Exception Handling

1- Try and Except

2-Finally

3-Unit Testing 



In [1]:
def add(n1,n2):
    print(n1+n2)

In [2]:
add(10,20)

30


In [3]:
number1 = 10

In [4]:
number2 = input("Please provide a number: ")

Please provide a number:  20


In [5]:
#it'll be get a error. because number2 is string. 
add(number1,number2)


TypeError: unsupported operand type(s) for +: 'int' and 'str'

### try and except
The basic terminology and syntax used to handle errors in Python are the try and except statements. The code which can cause an exception to occur is put in the try block and the handling of the exception is then implemented in the except block of code. The syntax follows:


try:
      
      You do your operations here...
   
   ...
   
except ExceptionI:

       If there is ExceptionI, then execute this block.
except ExceptionII:

       If there is ExceptionII, then execute this block.
  
  ...

else:

       If there is no exception then execute this block. 

We can also just check for any exception with just using except: To get a better understanding of all this let's check out an example: We will look at some code that opens and writes a file:

In [6]:
try:
    # WANT TO ATTEMPT THIS CODE
    # MAY HAVE AN ERROR
    result = 10 + 10
except:
    print("Hey it looks like you aren't adding correctly!")

In [7]:
result

20

In [10]:
try:
    # WANT TO ATTEMPT THIS CODE
    # MAY HAVE AN ERROR
    result = 10 + "10"
except:
    print("Hey it looks like you aren't adding correctly!")

Hey it looks like you aren't adding correctly!


In [11]:
try:
    # WANT TO ATTEMPT THIS CODE
    # MAY HAVE AN ERROR
    result = 10 + 10
except:
    print("Hey it looks like you aren't adding correctly!")
else:
    print("Add wnt well!")
    print(result)

Add wnt well!
20


### finally
The finally: block of code will always be run regardless if there was an exception in the try code block. The syntax is:

try:

    Code block here
   
   ...
   
    Due to any exception, this code may be skipped!

finally:

    This code block would always be executed.


In [12]:
try:
    f = open('testfile','w')
    f.write('Write a test line')
except TypeError:
    # This will only check for an TypeError exception and then execute this print statement
    print("There was a type error")
except OSError:
    print('Hey you have an OS Error')
finally:
    print("I always run")

I always run


In [14]:
try:
    f = open('testfile','r')
    f.write('Write a test line')
except TypeError:
    print("There was a type error")
except OSError:
    print('Hey you have an OS Error')
finally:
    print("I always run")

Hey you have an OS Error
I always run


In [17]:
try:
    f = open('testfile','r')
    f.write('Write a test line')
except TypeError:
    print("There was a type error")
except:
    print('All other exceptions!')
finally:
    print("I always run")

All other exceptions!
I always run


In [27]:
def ask_for_int():
    try:
        result = int(input("Please provide number: "))
    except:
        print("Whoops! That is not a number")
    finally:
        print("End of try/except/finally")

In [28]:
ask_for_int()

Please provide number:  123


End of try/except/finally


In [29]:
ask_for_int()

Please provide number:  abc


Whoops! That is not a number
End of try/except/finally


In [30]:
ask_for_int()

Please provide number:  1a2


Whoops! That is not a number
End of try/except/finally


In [31]:
def ask_for_int():
    
    while True:
        try:
            result = int(input("Please provide number: "))
        except:
            print("Whoops! That is not a number")
            continue
        else:
            print("Yes thank you")
            break
        finally:
            print("End of try/except/finally")
            print("I will always run at the end!")

In [32]:
ask_for_int()

Please provide number:  36


Yes thank you
End of try/except/finally
I will always run at the end!


In [33]:
ask_for_int()

Please provide number:  abc


Whoops! That is not a number
End of try/except/finally
I will always run at the end!


Please provide number:  2ş


Whoops! That is not a number
End of try/except/finally
I will always run at the end!


Please provide number:  23


Yes thank you
End of try/except/finally
I will always run at the end!


In [34]:
x = 5
y = 0
try:
    z = x/y
except ZeroDivisionError:
    print("Can't divide by Zero!")
finally:
    print('All Done!')

Can't divide by Zero!
All Done!


In [35]:
def ask():
    
    while True:
        try:
            n = int(input('Input an integer: '))
        except:
            print('An error occurred! Please try again!')
            continue
        else:
            break
            
        
    print('Thank you, your number squared is: ',n**2)

In [36]:
ask()

Input an integer:  23


Thank you, your number squared is:  529


In [38]:
ask()

Input an integer:  lk


An error occurred! Please try again!


Input an integer:  ab


An error occurred! Please try again!


Input an integer:  16


Thank you, your number squared is:  256


## Unit Testing

### Testing tools
There are dozens of good testing libraries out there. Most are third-party packages that require an install, such as:

pylint /
pyflakes /
pep8

These are simple tools that merely look at your code, and they'll tell you if there are style issues or simple problems like variable names being called before assignment.

A far better way to test your code is to write tests that send sample data to your program, and compare what's returned to a desired outcome.
Two such tools are available from the standard library:

unittest /
doctest

Let's look at pylint first, then we'll do some heavier lifting with unittest.

https://docs.pytest.org/en/stable/

In [1]:
%%writefile simple1.py
a = 1
b = 2
print(a)
print(B)

Writing simple1.py


In [2]:
! pylint simple1.py

************* Module simple1
simple1.py:1:0: C0114: Missing module docstring (missing-module-docstring)
simple1.py:1:0: C0103: Constant name "a" doesn't conform to UPPER_CASE naming style (invalid-name)
simple1.py:2:0: C0103: Constant name "b" doesn't conform to UPPER_CASE naming style (invalid-name)
simple1.py:4:6: E0602: Undefined variable 'B' (undefined-variable)

----------------------------------------------------------------------

Your code has been rated at -10.00/10 (previous run: -12.50/10, +2.50)





No config file found, using default configuration

Pylint first lists some styling issues - it would like to see an extra newline at the end, modules and function definitions should have descriptive docstrings, and single characters are a poor choice for variable names.

More importantly, however, pylint identified an error in the program - a variable called before assignment. This needs fixing.

Note that pylint scored our program a negative 12.5 out of 10. Let's try to improve that!

In [3]:
%%writefile simple1.py
"""
A very simple script.
"""

def myfunc():
    """
    An extremely simple function.
    """
    first = 1
    second = 2
    print(first)
    print(second)

myfunc()

Overwriting simple1.py


In [4]:
! pylint simple1.py


----------------------------------------------------------------------

Your code has been rated at 10.00/10 (previous run: -10.00/10, +20.00)





No config file found, using default configuration

Much better! Our score climbed to 8.33 out of 10. Unfortunately, the final newline has to do with how jupyter writes to a file, and there's not much we can do about that here. Still, pylint helped us troubleshoot some of our problems. But what if the problem was more complex?

In [5]:
%%writefile simple2.py
"""
A very simple script.
"""

def myfunc():
    """
    An extremely simple function.
    """
    first = 1
    second = 2
    print(first)
    print('second')

myfunc()

Writing simple2.py


In [6]:
! pylint simple2.py

************* Module simple2
simple2.py:10:4: W0612: Unused variable 'second' (unused-variable)

-----------------------------------

Your code has been rated at 8.33/10





No config file found, using default configuration

pylint tells us there's an unused variable in line 10, but it doesn't know that we might get an unexpected output from line 12! For this we need a more robust set of tools. That's where unittest comes in.

### unittest
unittest lets you write your own test programs. The goal is to send a specific set of data to your program, and analyze the returned results against an expected result.

Let's generate a simple script that capitalizes words in a given string. We'll call it cap.py.

In [7]:
%%writefile cap.py
def cap_text(text):
    return text.capitalize()

Writing cap.py


In [8]:
%%writefile test_cap.py
import unittest
import cap

class TestCap(unittest.TestCase):
    
    def test_one_word(self):
        text = 'python'
        result = cap.cap_text(text)
        self.assertEqual(result, 'Python')
        
    def test_multiple_words(self):
        text = 'monty python'
        result = cap.cap_text(text)
        self.assertEqual(result, 'Monty Python')
        
if __name__ == '__main__':
    unittest.main()

Writing test_cap.py


In [9]:
! python test_cap.py

F.
FAIL: test_multiple_words (__main__.TestCap)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\User\Desktop\pandas\test_cap.py", line 14, in test_multiple_words
    self.assertEqual(result, 'Monty Python')
AssertionError: 'Monty python' != 'Monty Python'
- Monty python
?       ^
+ Monty Python
?       ^


----------------------------------------------------------------------
Ran 2 tests in 0.001s

FAILED (failures=1)


What happened? It turns out that the .capitalize() method only capitalizes the first letter of the first word in a string. Doing a little research on string methods, we find that .title() might give us what we want.

In [10]:
%%writefile cap.py
def cap_text(text):
    return text.title()  # replace .capitalize() with .title()

Overwriting cap.py


In [11]:
! python test_cap.py

..
----------------------------------------------------------------------
Ran 2 tests in 0.000s

OK


Hey, it passed! But have we tested all cases? Let's add another test to test_cap.py to see if it handles words with apostrophes, like don't.

In a text editor this would be easy, but in Jupyter we have to start from scratch.

In [12]:
%%writefile test_cap.py
import unittest
import cap

class TestCap(unittest.TestCase):
    
    def test_one_word(self):
        text = 'python'
        result = cap.cap_text(text)
        self.assertEqual(result, 'Python')
        
    def test_multiple_words(self):
        text = 'monty python'
        result = cap.cap_text(text)
        self.assertEqual(result, 'Monty Python')
        
    def test_with_apostrophes(self):
        text = "monty python's flying circus"
        result = cap.cap_text(text)
        self.assertEqual(result, "Monty Python's Flying Circus")
        
if __name__ == '__main__':
    unittest.main()

Overwriting test_cap.py


In [13]:
! python test_cap.py

..F
FAIL: test_with_apostrophes (__main__.TestCap)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\User\Desktop\pandas\test_cap.py", line 19, in test_with_apostrophes
    self.assertEqual(result, "Monty Python's Flying Circus")
AssertionError: "Monty Python'S Flying Circus" != "Monty Python's Flying Circus"
- Monty Python'S Flying Circus
?              ^
+ Monty Python's Flying Circus
?              ^


----------------------------------------------------------------------
Ran 3 tests in 0.001s

FAILED (failures=1)
